pkg/apiserver/apiserver.go

```diff
@@ -0,0 +1,561 @@
+	return nil
+}
+
+func (app *uploadAPIApp) startTLS() error {
+
+	errors := make(chan error)
+
+	keyFile := filepath.Join(app.certsDirectory, "/key.pem")
+	certFile := filepath.Join(app.certsDirectory, "/cert.pem")
+	signingCertFile := filepath.Join(app.certsDirectory, "/signingCert.pem")
+	clientCAFile := filepath.Join(app.certsDirectory, "/clientCA.crt")
+
+	// Write the certs to disk
+	err := ioutil.WriteFile(clientCAFile, app.clientCABytes, 0600)
+	if err != nil {
+		return err
+	}
+
+	if len(app.requestHeaderClientCABytes) != 0 {
+		f, err := os.OpenFile(clientCAFile, os.O_APPEND|os.O_WRONLY, 0600)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+
+		_, err = f.Write(app.requestHeaderClientCABytes)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = ioutil.WriteFile(keyFile, app.keyBytes, 0600)
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(certFile, app.certBytes, 0600)
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(signingCertFile, app.signingCertBytes, 0600)
+	if err != nil {
+		return err
+	}
+
+	// create the client CA pool.
+	// This ensures we're talking to the k8s api server
+	pool, err := cert.NewPool(clientCAFile)
+	if err != nil {
+		return err
+	}
+
+	tlsConfig := &tls.Config{
+		ClientCAs:  pool,
+		ClientAuth: tls.RequestClientCert,
+	}
+	tlsConfig.BuildNameToCertificate()
+
+	go func() {
+		server := &http.Server{
+			Addr:      fmt.Sprintf("%s:%d", app.bindAddress, app.bindPort),
+			TLSConfig: tlsConfig,
+		}
+
+		errors <- server.ListenAndServeTLS(certFile, keyFile)
+	}()
+
+	// wait for server to exit
+	return <-errors
+}
+
+func (app *uploadAPIApp) uploadHandler(request *restful.Request, response *restful.Response) {
+
+	allowed, reason, err := app.authorizor.Authorize(request)
+	if err != nil {
+		glog.Error(err)
+		response.WriteHeader(http.StatusInternalServerError)
+		return
+	} else if !allowed {
+		glog.Infof("Rejected Request: %s", reason)
+		response.WriteErrorString(http.StatusUnauthorized, reason)
+		return
+	}
+
+	namespace := request.PathParameter("namespace")
+	defer request.Request.Body.Close()
+	body, err := ioutil.ReadAll(request.Request.Body)
+	if err != nil {
+		glog.Error(err)
+		response.WriteError(http.StatusBadRequest, err)
+		return
+	}
+
+	uploadToken := &cdiv1.UploadToken{}
+	err = json.Unmarshal(body, uploadToken)
+	if err != nil {
+		glog.Error(err)
+		response.WriteError(http.StatusBadRequest, err)
+		return
+	}
+
+	pvcName := uploadToken.Spec.PvcName
+	pvc, err := app.client.CoreV1().PersistentVolumeClaims(namespace).Get(pvcName, metav1.GetOptions{})
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			glog.Infof("Rejecting request for PVC %s that doesn't exist", pvcName)
+			response.WriteError(http.StatusBadRequest, err)
+			return
+		}
+		glog.Error(err)
+		response.WriteError(http.StatusInternalServerError, err)
+		return
+	}
+
+	if err = controller.UploadPossibleForPVC(pvc); err != nil {
+		response.WriteError(http.StatusServiceUnavailable, err)
```

Comment:
should there be a return statement here to prevent the token from being generated?

Reply:
whoops!  yeah definitely.